doc/neps/nep-0018-array-function-protocol.rst

```diff
@@ -250,87 +304,113 @@ an ``__array_function__`` method that is identical to
-
-        ...  # continue with the definition of concatenate
-
-The list of objects passed to ``relevant_arguments`` are those that should
-be inspected for ``__array_function__`` implementations.
-
-Our microbenchmark results show that a pure Python implementation of
-``try_array_function_override`` adds approximately 2-4 microseconds of
-overhead to each function call without any overloaded arguments.
-This is acceptable for functions implemented in Python but probably too
-slow for functions written in C. Fortunately, we expect significantly less
-overhead with a C implementation of ``try_array_function_override``.
-
-A more succinct alternative would be to write these overloads with a decorator
-that builds overloaded functions automatically. Hypothetically, this might even
-directly parse Python 3 type annotations, e.g., perhaps
+In most cases, these functions should written using a shared ``dispatch``
+decorator, which also associates dispatcher functions:
 
 .. code:: python
 
-    @overload_for_array_function
-    def broadcast_to(array: ArrayLike
-                     shape: Tuple[int, ...],
-                     subok: bool = False):
-        ...  # continue with the definition of broadcast_to
-
-The decorator ``overload_for_array_function`` would be written in terms
-of ``try_array_function_override``, but would also need some level of magic
-for (1) access to the wrapper function (``np.broadcast_to``) for passing into
-``__array_function__`` implementations and (2) dynamic code generation
-resembling the `decorator library <https://github.com/micheles/decorator>`_ 
-to automatically write an overloaded function like the manually written
-implemenations above with the exact same signature as the original.
-Unfortunately, using the ``inspect`` module instead of code generation would
-probably be too slow: our prototype implementation adds ~15 microseconds of
-overhead.
-
-We like the idea of writing overloads with minimal syntax, but dynamic
-code generation also has potential downsides, such as slower import times, less
-transparent code and added difficulty for static analysis tools. It's not clear
-that tradeoffs would be worth it, especially because functions with complex
-signatures like ``np.einsum`` would assuredly still need to invoke
-``try_array_function_override`` directly.
-
-So we don't propose adding such a decorator yet, but it's something worth
-considering for the future.
+    def dispatch(dispatcher):
+        """Wrap a function for dispatch with the __array_function__ protocol."""
+        def decorator(func):
+            @functools.wraps(func)
+            def new_func(*args, **kwargs):
+                relevant_arguments = dispatcher(*args, **kwargs)
+                success, value = try_array_function_override(
+                    new_func, relevant_arguments, args, kwargs)
+                if success:
+                    return value
+                return func(*args, **kwargs)
+            return new_func
```

Comment:
To attach the dispatcher, wouldn't one just do
```
new_func.dispatcher = dispatcher
return new_func
```

Reply:
This doesn't work for "builtin" functions written in C like `np.concatenate`:
```
>>> np.concatenate.foo = 1
---------------------------------------------------------------------------
AttributeError                            Traceback (most recent call last)
<ipython-input-2-a0d69ae14b06> in <module>()
----> 1 np.concatenate.foo = 1

AttributeError: 'builtin_function_or_method' object has no attribute 'foo'
```

Reply:
But in `__array_ufunc__`, aren't we passing on a wrapped `concatenate`? One can definitely assign to the wrapped function which would become the `np.concatenate` that people actually see:
```
def new_concat(*args, **kwargs):
    return np.concatenate(*args, **kwargs)

nc = functools.wraps(np.concatenate)(new_concat)
nc.dispatch = 'a'
```
(obviously, the same would be possible in C)

Reply:
It depends on how we choose to implement `np.concatenate` with dispatching. Potentially we could do it either way.

I'm sure it's also possible to define a normal Python function from C directly rather than a builtin. I don't know the trade-offs there.

Reply:
I think it would be wise to keep the wrapped and unwrapped versions separate, so that, e.g., `__array_function__` can directly call the unwrapped one. (I still hope to do the same for the ufuncs).

Reply:
And, indeed, for a C implementation one might as well build the python function in C; the API does seem like it wouldn't be that hard (base on having seen a developed interface, not direct experience).

Reply:
p.s. one of the points of the wrapping in the example is to have access to `dispatch` without actually having to pass that in; it therefore must always be possible to expose it!

Reply:
> p.s. one of the points of the wrapping in the example is to have access to dispatch without actually having to pass that in; it therefore must always be possible to expose it!

I'm not sure I follow here: what is "it"? :)

Reply:
`dispatch`: your wrapped function *must* be able to call `dispatch` and therefore has access to it, and must thus be able to expose it to the outside. Of course, this does rely on having the wrapped function be the one in the `numpy` namespace and identical to the one passed on in `__array_function__`.

Maybe most relevant to the NEP: maybe good to state a preference for C functions being wrapped, not having dispatch cooked inside.

Reply:
Sorry, yes, I meant the `dispatcher` used inside `new_func`